brave/src/main/java/brave/internal/recorder/MutableSpan.java

```diff
@@ -31,11 +38,27 @@
    * To reduce the amount of allocation, collocate annotations with tags in a pair-indexed list.
    * This will be (startTimestamp, value) for annotations and (key, value) for tags.
    */
-  final List<Object> pairs = new ArrayList<>(6); // assume 3 tags and no annotations
+  final ArrayList<Object> pairs;
```

Comment:
`ArrayList` will cost when the length extend happens. Linkedlist is always faster.

Reply:
https://github.com/google/error-prone/blob/0cc518144272ea6150c85e2cfb937c82a8d85223/docs/bugpattern/JdkObsolete.md#linkedlist so we will have to make an exception if benchmarks show you are correct. this is cool though.. internal detail will check it!

Reply:
I am always wondering, why linkedlist should not be used.

Reply:
The main criticism of LinkedLists is pointer chasing. When using an ArrayList, the memory layout is much denser and the memory access pattern is easier to predict by CPUs. With linked lists, the next elements are less likely in the same cache line which leads to expensive cache misses and main memory access.